Added Skip Logic: Matrix parameters cannot be empty arrays (#6140)
tektoncd/pipeline: docs/matrix.md

`PipelineTask` by substituting `Parameters` of type `"string"` in the underlying `Task`.
The names of the `Parameters` in the `Matrix` must match the names of the `Parameters`
in the underlying `Task` that they will be substituting.

> Note: The `Paramaters` of type `array` cannot be empty.

Reviewer: thank you @EmmaMunley for updating the doc, I appreciate it but this is not my expected behavior is. 

As a user, I would want to:

1. With a single matrix param - `skip` the pipelineTask if matrix param is an empty array. 
2. With multiple matrix params - If all params are empty array, follow (1) else create param combination with non-empty matrix params for the underlying pipelinesTask.

There is a distinction between an empty array and uninitialized array. Uninitialized array is treated as an error but empty array could have been chosen by design. 

@jerop thoughts?

Reviewer 2: @pritidesai do you have an example or use case for why an empty array should be valid? tried to think of something but couldn't come up with any because I assume that a value has to be passed to the taskrun for that param

also, if there's no non-empty array, I'd expect `failure`, instead of `skipping`, because we can't create a taskrun from that `pipelinetask`

Reviewer: thank you @jerop!

We had a discussion in the [API WG](https://docs.google.com/document/d/1XTN44T_AzAXDRHS6relY-Au2DzoaHSu2MpT0UkH01TA/edit#heading=h.ah8u3qqpv18t) with @EmmaMunley and @afrittoli on this topic.

Use case: A pipeline is generating SBOMs using many different tools and uploading them to an object store. Here, a list of SBOMs is a matrix param. It is possible that none of the tools produced SBOM for some valid reason. As a pipeline author, I would like to continue with the rest of the pipeline instead of failing the entire run. Generating these SBOMs could be just a part of one pipeline with other tasks. As a pipeline author,

One more use case I documented in the WG notes with the list of images. As a pipeline author, test the list of images which were published based on some criteria. It is possible that there were no images published. In that case, the pipeline will not run tests but will continue running the rest of the tasks of why the images failed to meet the required criteria.

Failure is fatal and will cause the entire pipeline to fail. The pipeline could have spent hours on a list of tasks. 

>also, if there's no non-empty array, I'd expect failure, instead of skipping, because we can't create a taskrun from that pipelinetask

Please correct me if I am wrong but skipping anyways does not create a `taskRun`.

Also, this might need a little more discussion.

Now, the whole purpose of `matrix` is to generate exhaustive list of combinations for a given set of parameters.

In case of a matrix with one param and that one param being empty, `matrix` generates 0 combination. So, the natural choice is to skip the pipelineTask `my-matrix-task` because pipeline controller does not have enough information to decide whether to create one taskRun or 10 taskRuns and what should be the value of that param `param1`. Now, this param `param1` could be  associated to optional or required param of the task. We could run the underlying task once with a default value of that param. And in case of a required param, fail the pipelineRun since the required param is not initialized. But its contracting `matrix` contract of initiating taskRuns based on the number of combinations of matrix params.  

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[]"
```

Therefore, the best way forward could be skip creating a taskRun and therefore declare that pipelineTask as skipped. 

But for a matrix with more than one params and one of the params is an empty array:

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[a, b, c]"
        - name: param2
          value: "[]"
```

This `matrix` can result in three combinations with `param2` as an empty string for all three combinations or pick a default value from the task for this param. Or even when a `matrix` resulted in three combinations, skip the pipelineTask `my-matrix-task` because one of the matrix params was an empty array.

And, both matrix params set to an empty array:

```
  - name: my-matrix-task
    matrix:
      params:
        - name: param1
          value: "[]"
        - name: param2
          value: "[]"
```

This article has a few examples on `matrix` configuration based on some use cases: https://ncorti.com/blog/howto-github-actions-build-matrix

To summarize, `matrix` must not cause a pipelineRun to fail when a matrix param is set to an empty array. In case of a `matrix` without `include` from [TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md), skip if any one of the matrix param is an empty array.

I am putting on hold until we can come to conclusion (hope very soon)

/hold

Reviewer 3: Thanks, @pritidesai - it seems to me that there are two levels to the issue, and I think the decision for both should be consistent regardless of the number of matrixed params and number of zero length arrays

1.  Are empty arrays in a matrix valid? i.e. `PipelineRun` continues running or not in case of empty arrays in matrix. Because we want to support dynamic arrays in future, I think it would be preferable *not* to fail, i.e. consider an empty array a valid alternative
2. Does the `PipelineRun` controller create `TaskRuns` in case of empty arrays in matrix? If one or more dimensions are empty the controller could still create `TaskRuns` without the specific parameter and let the `TaskRun` controller decide whether to run it (if there is a default set) or not. This could be a desirable behaviour in some cases, but I feel like it could be confusing for users, so my recommendation would be to skip if any of the matrixed arrays are empty

In case of skip, we should use apply the same logic that we apply for `when` expressions - the default skip logic would be to continue with following tasks, unless the skip logic is set to branch.

I'm not sure if we support results from a task in a matrix today, but if we do, the resulting array of results should be empty, so that the following tasks could decide how to react based on the size. 
I believe we don't support the `${#array}` syntax yet to evaluate the size of an array, I believe that would be a nice addition for these cases.

Reviewer 2: thanks for sharing the use cases and example scenarios @pritidesai!

agree with @afrittoli about consistency - let's start by skipping the pipelinetask if any param in a matrix is empty...we can revisit the idea of fanning out using non-empty params only (and ignoring the empty params) in the future so that we can keep things simple for now, what do you think @pritidesai?

@afrittoli - we do not yet support results from a matrixed pipelinetask - [TEP-0090](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md#results-from-fanned-out-pipelinetasks )

Reviewer 2: Discussed with @pritidesai at API WG on 3/6 and agreed to move forward with @afrittoli's suggestion that we skip the PipelineTask if there's an empty array in the Matrix -- cc @EmmaMunley

Reviewer: thanks a bunch @jerop @afrittoli and others for all the input! Very excited to see we have come to the conclusion 🎉 

To summarize:

The initial proposal was to `fail` the `pipelineRun` when a `matrix.params` has an empty array. My initial suggestion was to `skip` in two scenarios (1) a single `matrix.params` set to empty array (2) all specified `matrix.params` are set to empty array but continue execution in third scenario (3) all specified `matrix.params` are not empty i.e. some of them are initialized to a valid array. @afrittoli suggested to keep it consistent and `skip` in all three scenarios (thank you @afrittoli for the great suggestion).

@EmmaMunley please feel free to ping me if you have any questions. I will be very happy to guide you further.

The `Matrix` will take `Parameters` of type `"array"` only, which will be supplied to the
`PipelineTask` by substituting `Parameters` of type `"string"` in the underlying `Task`.

> A `Parameter`that contains an empty array is considered invalid. The names of the
> `Parameters` in the `Matrix` must match the names of the `Parameters`

Reviewer 2: it is invalid implies that it'd fail, let's make it clear that it'd be skipped

```suggestion
If the `Matrix` has an empty array `Parameter`, then the `PipelineTask` will be skipped.
The names of the `Parameters` in the `Matrix` must match the names of the `Parameters`
```
